### PR TITLE
DAOS-16170 cart: do not release completed RPC reference  repeatedly

### PR DESCRIPTION
**Pull request by Nasf-Fan**
For collective RPC, when handle failure cases during crt_req_send(),
its reference may has been released via crt_rpc_complete_and_unlock()
that is triggered by crt_corpc_complete(). Under such case, we should
check whether the RPC is completed or not before calling RPC_DECREF()
to avoid releasing the RPC reference repeatedly.

The patch also initializes some local variable for CHK RPC to avoid
accessing invalid DRAM when handle failed collective CHK RPC.

Some enhancement for CR test logic.

Test-tag: test_daos_cat_recov_core
Allow-unstable-test: true

Signed-off-by: Fan Yong <fan.yong@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by Nasf-Fan** [comment]
With the patch, CR tests passed on CI:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15476/5/artifact/Functional%20Hardware%20Medium%20Verbs%20Provider/recovery/cat_recov_core.py/

But seems only CR was tests this cycle, let's tigger another CI tests for others.

File: src/cart/crt_rpc.c [hunk 1]
```diff
@@ -1532,7 +1532,7 @@ crt_req_send(crt_rpc_t *req, crt_cb_t complete_cb, void *arg)
 			/* failure already reported through complete cb */
 			if (complete_cb != NULL)
 				rc = 0;
-		} else {
+		} else if (!crt_rpc_completed(rpc_priv)) {
```
Comment (frostedcmos) [inline src/cart/crt_rpc.c:1535]: Copying here the feedback from 2.6 branch:

Can you explain how an rpc can be validly completed at this point?

From my view, to complete, we would need to send it via 1515 rc = crt_req_send_internal(rpc_priv); , but that call should either a return rc on error, or send the rpc and invoke a completion cb. Your check here implies that somehow both can happen.

One note to make here, we only used crt_rpc_completed() to detect bugs in the ref counting logic or elsewhere before. I don't think we should start using this function to decide ref counting, and instead should fix underlying ref count problem instead

**Review by frostedcmos — CHANGES_REQUESTED** [review]

File: src/cart/crt_rpc.c (hunk 1 above)
Comment (frostedcmos) [reply src/cart/crt_rpc.c:1535]: as discussed in 2.6 pr, this is fine for now. cart team will refine rpc completion logic later.

**Review by frostedcmos — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by gnailzenh — APPROVED** [review]

File: src/cart/crt_rpc.c (hunk 1 above)
Comment (liuxuezhao) [reply src/cart/crt_rpc.c:1535]: I did some investigation about the corpc reference, pushed another small PR -
https://github.com/daos-stack/daos/pull/15572
The refcount possibly can be fourthly refined later, the PR15572 just try to refine PR15476. 

@frostedcmos the refcount handling in crt_corpc_complete() looks correct, see below debug log for the refcount trace for the root node of a bcast, the same last number in the "()" means the corresponding addref and decref. The refcount handling looks fine, but the CRT_RPC_FLAG_CO_FAILOUT fail handling looks caused the issue.

12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:446 crt_corpc_req_create() corpc 0x7f329c46a740 created
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:783 crt_req_addref() rank 0 rpc 0x7f329c46a740, addref to 2                   (1) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1482 crt_req_send() rank 0 rpc 0x7f329c46a740, addref to 3                    (2) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:787 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 4            (3) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:889 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 5            (4) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] corpc ERR  src/cart/crt_corpc.c:889 crt_corpc_req_hdlr() rank 0 rpc 0x7f329c46a740, addref to 6            (5) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1854 crt_rpc_common_hdlr() rank 0 rpc 0x7f329c46a740, addref to 7             (6) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:1544 crt_req_send() rank 0 rpc 0x7f329c46a740, decref to 6                    (2) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1789 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, addref to 7                  (7) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1798 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, decref to 6                  (7) 
12/07-03:56:02.31 wolf-419 DAOS[1220393/0/74] rpc  ERR  src/cart/crt_rpc.c:1804 crt_handle_rpc() rank 0 rpc 0x7f329c46a740, decref to 5                  (6) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:768 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 4           (4) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:755 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, addref to 5           (8) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] rpc  ERR  src/cart/crt_context.c:550 crt_rpc_complete_and_unlock() rank 0 rpc 0x7f329c46a740, decref to 4 (creat)
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:576 crt_corpc_complete() rank 0 rpc 0x7f329c46a740, decref to 3             (3) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:763 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 2           (8) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/5] corpc ERR  src/cart/crt_corpc.c:768 crt_corpc_reply_hdlr() rank 0 rpc 0x7f329c46a740, decref to 1           (5) 
12/07-03:56:02.32 wolf-419 DAOS[1220393/0/73] rpc  ERR  src/cart/crt_rpc.c:789 crt_req_decref() rank 0 rpc 0x7f329c46a740, decref to 0                   (1)

**Pull request merged**